File: src/sph/core/world.cpp
#include "world.h"
#ifdef SPH_ENABLE_HASH2D
#include "../debug_gpu.hpp"
#endif

namespace sph {

float floatRand() {
    return static_cast<float>(rand()) / static_cast<float>(RAND_MAX);
}

GridMap::GridMap(float width, float height, float radius)
    : width(static_cast<int>(width)),
      height(static_cast<int>(height)),
      chunkRange(radius) {
    numChunkWidth = static_cast<int>(std::ceil(this->width / this->chunkRange)) + 1;
    numChunkHeight = static_cast<int>(std::ceil(this->height / this->chunkRange)) + 1;
    numChunk = numChunkWidth * numChunkHeight;
    chunks.assign(numChunk, std::vector<int>());
}

std::vector<int> GridMap::getChunk(int chunkX, int chunkY) const {
    return chunks[chunkY * numChunkWidth + chunkX];
}

void GridMap::registerTarget(int target, float x, float y) {
    int chunkX = static_cast<int>(x / chunkRange);
    int chunkY = static_cast<int>(y / chunkRange);
    if (chunkX < 0 || chunkX >= numChunkWidth) return;
    if (chunkY < 0 || chunkY >= numChunkHeight) return;
    chunks[chunkY * numChunkWidth + chunkX].push_back(target);
}

void GridMap::unregisterAll() {
    for (int i = 0; i < numChunk; ++i) {
        chunks[i].clear();
    }
}

std::vector<int> GridMap::findNeighborhood(float x, float y, float radius) const {
    std::vector<int> out;
    int centerChunkX = static_cast<int>(x / chunkRange);
    int centerChunkY = static_cast<int>(y / chunkRange);
    int radiusChunk = static_cast<int>(std::ceil(radius / chunkRange));
    int minChunkX = centerChunkX - radiusChunk;
    int maxChunkX = centerChunkX + radiusChunk;
    int minChunkY = centerChunkY - radiusChunk;
    int maxChunkY = centerChunkY + radiusChunk;

    for (int x1 = minChunkX; x1 <= maxChunkX; ++x1) {
        for (int y1 = minChunkY; y1 <= maxChunkY; ++y1) {
            if (x1 < 0 || x1 >= numChunkWidth) continue;
            if (y1 < 0 || y1 >= numChunkHeight) continue;
            auto chunk = getChunk(x1, y1);
            out.insert(out.end(), chunk.begin(), chunk.end());
        }
    }
    return out;
}

World::World(const WorldConfig& config)
    : numParticle(config.numParticles),
      gravity(config.gravity),
      worldSize{config.worldWidth, config.worldHeight},
      collisionDamping(config.collisionDamping),
      smoothingRadius(config.smoothingRadius),
      targetDensity(config.targetDensity),
      pressureMultiplier(config.pressureMultiplier),
      delta(config.delta),
      drag(config.drag),
      forcePoint{{0,0},0,0},
      gridmap(worldSize[0], worldSize[1], smoothingRadius)
{
    pos.resize(numParticle);
    predpos.resize(numParticle);
    vel.resize(numParticle);
    density.assign(numParticle, 0.0f);
    pressureAccelerations.resize(numParticle);
    interactionForce.resize(numParticle);
    color.resize(numParticle);
    mass.assign(numParticle, 1.0f);

    for (int i = 0; i < numParticle; ++i) {
        int a = static_cast<int>(std::sqrt(static_cast<float>(numParticle)));
        int row = i / a;
        int col = i % a;
        pos[i][0] = (col / static_cast<float>(a)) * worldSize[0];
        pos[i][1] = (row / static_cast<float>(a)) * worldSize[1];
        predpos[i][0] = 0.0f;
        predpos[i][1] = 0.0f;
        vel[i][0] = 0.0f;
        vel[i][1] = 0.0f;
        pressureAccelerations[i] = {0.0f, 0.0f};
        interactionForce[i] = {0.0f, 0.0f};
        color[i] = {255,255,255};
    }

    iterator.resize(numParticle);
    for (int i = 0; i < numParticle; ++i) iterator[i] = i;
}

void World::setInteractionForce(float posX, float posY, float radius, float strength) {
    forcePoint = {{posX, posY}, radius, strength};
}

void World::deleteInteractionForce() {
    forcePoint = {{0,0},0,0};
}

float World::getWorldWidth() const { return worldSize[0]; }
float World::getWorldHeight() const { return worldSize[1]; }

void World::update(float deltaTime) {
    predictedPos(deltaTime);
    gridmap.unregisterAll();

    std::vector<int> v = iterator;
    std::for_each(std::execution::seq, v.begin(), v.end(), [&](int idx){
        gridmap.registerTarget(idx, pos[idx][0], pos[idx][1]);
    });

    querysize.resize(numParticle);
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){
        querysize[idx] = gridmap.findNeighborhood(pos[idx][0], pos[idx][1], smoothingRadius);
    });

    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ updateDensity(idx); });
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ updatePressureForce(idx); });
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ updateInteractionForce(idx); });

    updatePosition(deltaTime);
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ fixPositionFromWorldSize(idx); });
    updateColor();
}

void World::stepGPU(float deltaTime) {
#ifdef SPH_ENABLE_HASH2D
    predictedPos(deltaTime);
    gridmap.unregisterAll();
    std::vector<int> v = iterator;
    std::for_each(std::execution::seq, v.begin(), v.end(), [&](int idx){
        gridmap.registerTarget(idx, pos[idx][0], pos[idx][1]);
    });

    allocateDeviceBuffers();

    std::vector<float2> hPos(numParticle);
    for (int i = 0; i < numParticle; ++i) {
        hPos[i] = make_float2(pos[i][0], pos[i][1]);
    }
    CUDA_TRY(cudaMemcpy(grid.particles.pos,
                        hPos.data(),
                        sizeof(float2) * numParticle,
                        cudaMemcpyHostToDevice));

    grid.build(numParticle);
    grid.findNeighbors(numParticle, smoothingRadius, d_neighbors, d_counts);
    CUDA_TRY(cudaDeviceSynchronize());

    std::vector<uint32_t> hCounts(numParticle);
    std::vector<uint32_t> hNeighbors(numParticle * MAX_NEIGHBORS);
    CUDA_TRY(cudaMemcpy(hCounts.data(),
                        d_counts,
                        sizeof(uint32_t) * numParticle,
                        cudaMemcpyDeviceToHost));
    CUDA_TRY(cudaMemcpy(hNeighbors.data(),
                        d_neighbors,
                        sizeof(uint32_t) * numParticle * MAX_NEIGHBORS,
                        cudaMemcpyDeviceToHost));

    querysize.assign(numParticle, {});
    for (int i = 0; i < numParticle; ++i) {
        uint32_t c = hCounts[i];
        querysize[i].resize(c);
        for (uint32_t j = 0; j < c; ++j) {
            querysize[i][j] = static_cast<int>(hNeighbors[i * MAX_NEIGHBORS + j]);
        }
    }

    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ updateDensity(idx); });
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ updatePressureForce(idx); });
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ updateInteractionForce(idx); });

    updatePosition(deltaTime);
    std::for_each(std::execution::par_unseq, v.begin(), v.end(), [&](int idx){ fixPositionFromWorldSize(idx); });
    updateColor();
#else
    update(deltaTime);
#endif
}

void World::predictedPos(float deltaTime) {
    for (int i = 0; i < numParticle; ++i) {
        vel[i][0] += 0.0f;
        vel[i][1] += mass[i] * gravity * deltaTime;
        predpos[i][0] = pos[i][0] + vel[i][0] * 1.0f / 120.0f;
        predpos[i][1] = pos[i][1] + vel[i][1] * 1.0f / 120.0f;
    }
}

void World::updateDensity(int particleIndex) { density[particleIndex] = calcDensity(particleIndex); }

void World::updatePressureForce(int particleIndex) {
    float pressureForce[] = {0,0};
    calcPressureForce(pressureForce, particleIndex);
    pressureAccelerations[particleIndex][0] = pressureForce[0] / (density[particleIndex] + delta);
    pressureAccelerations[particleIndex][1] = pressureForce[1] / (density[particleIndex] + delta);
}

void World::updateInteractionForce(int i) {
    float outForce[] = {0,0};
    calcInteractionForce(outForce, i);
    interactionForce[i][0] = outForce[0];
    interactionForce[i][1] = outForce[1];
}

void World::updatePosition(float deltaTime) {
    for (int i = 0; i < numParticle; ++i) {
        vel[i][0] += (pressureAccelerations[i][0] + interactionForce[i][0]) * deltaTime;
        vel[i][1] += (pressureAccelerations[i][1] + interactionForce[i][1]) * deltaTime;
        pos[i][0] += vel[i][0] * deltaTime;
        pos[i][1] += vel[i][1] * deltaTime;
        vel[i][0] *= drag;
        vel[i][1] *= drag;
    }
}

void World::fixPositionFromWorldSize(int i) {
    float x = pos[i][0];
    float y = pos[i][1];
    float velX = vel[i][0];
    float velY = vel[i][1];
    int w = static_cast<int>(worldSize[0]);
    int h = static_cast<int>(worldSize[1]);
    if (x < 0) { pos[i][0] = 0; vel[i][0] = -velX * collisionDamping; }
    if (w < x) { pos[i][0] = w; vel[i][0] = -velX * collisionDamping; }
    if (y < 0) { pos[i][1] = 0; vel[i][1] = -velY * collisionDamping; }
    if (h < y) { pos[i][1] = h; vel[i][1] = -velY * collisionDamping; }
}

void World::updateColor() {
    std::vector<float> speeds(numParticle);
    float minSpeed = FLT_MAX;
    float maxSpeed = 0.0f;
    int color1[3] = {0,0,255};
    int color2[3] = {255,0,0};
    for (int i = 0; i < numParticle; ++i) {
        float speed = std::sqrt(vel[i][0]*vel[i][0] + vel[i][1]*vel[i][1]);
        if (minSpeed > speed) minSpeed = speed;
        if (maxSpeed < speed) maxSpeed = speed;
        speeds[i] = speed;
    }
    for (int i = 0; i < numParticle; ++i) {
        float normSpeed = (speeds[i] - minSpeed) / (maxSpeed - minSpeed);
        normSpeed = std::clamp(normSpeed, 0.0f, 1.0f);
        uint8_t byteVel = static_cast<uint8_t>(normSpeed * 255.0f);
        color[i][0] = ((255 - byteVel) * color1[0] + byteVel * color2[0]) / 255;
        color[i][1] = ((255 - byteVel) * color1[1] + byteVel * color2[1]) / 255;
        color[i][2] = ((255 - byteVel) * color1[2] + byteVel * color2[2]) / 255;
    }
}

float World::calcDensity(int particleIndex) {
    float densityVal = 0.0f;
    auto otherIndexes = querysize[particleIndex];
    std::vector<float> distances(otherIndexes.size());
    for (size_t idx = 0; idx < otherIndexes.size(); ++idx) {
        int j = otherIndexes[idx];
        float dx = predpos[j][0] - predpos[particleIndex][0];
        float dy = predpos[j][1] - predpos[particleIndex][1];
        distances[idx] = std::sqrt(dx*dx + dy*dy);
    }
    std::vector<float> influences(otherIndexes.size());
    sph::calcSmoothingKernelCUDA(distances.data(), influences.data(), smoothingRadius, static_cast<int>(otherIndexes.size()));
    for (size_t idx = 0; idx < otherIndexes.size(); ++idx) {
        int j = otherIndexes[idx];
        densityVal += mass[j] * influences[idx];
    }
    return densityVal;
}

void World::calcPressureForce(float pressureForce[], int particleIndex) {
    auto otherIndexes = querysize[particleIndex];
    for (int otherIndex : otherIndexes) {
        if (particleIndex == otherIndex) continue;
        float offsetX = pos[otherIndex][0] - pos[particleIndex][0];
        float offsetY = pos[otherIndex][1] - pos[particleIndex][1];
        float dist = std::sqrt(offsetX*offsetX + offsetY*offsetY);
        if (dist > smoothingRadius) continue;
        float dirX = 0.0f;
        float dirY = 0.0f;
        if (dist <= FLT_EPSILON) {
            dirX = floatRand() - 0.5f;
            dirY = floatRand() - 0.5f;
        } else {
            dirX = offsetX / dist;
            dirY = offsetY / dist;
        }
        float slope = sph::calcSmoothingKernelDerivative(dist, smoothingRadius);
        float otherDensity = density[otherIndex];
        float sharedPressure = calcSharedPressure(otherDensity, density[particleIndex]);
        float a = sharedPressure * slope * mass[otherIndex] / (otherDensity + delta);
        pressureForce[0] += dirX * a;
        pressureForce[1] += dirY * a;
    }
}

float World::calcSharedPressure(float densityLeft, float densityRight) const {
    float pressureLeft = convertDensityToPressure(densityLeft);
    float pressureRight = convertDensityToPressure(densityRight);
    return (pressureLeft + pressureRight) * 0.5f;
}

float World::convertDensityToPressure(float densityVal) const {
    float densityError = densityVal - targetDensity;
    float pressure = densityError * pressureMultiplier;
    return pressure;
}

void World::calcInteractionForce(float outForce[], int particleIndex) const {
    outForce[0] = 0.0f;
    outForce[1] = 0.0f;
    ForcePoint p = forcePoint;
    float offsetX = p.pos[0] - pos[particleIndex][0];
    float offsetY = p.pos[1] - pos[particleIndex][1];
    float sqrDst = offsetX*offsetX + offsetY*offsetY;
    if (!(sqrDst < p.radius * p.radius)) return;
    float dirToForcePosX = 0.0f;
    float dirToForcePosY = 0.0f;
    float dist = std::sqrt(sqrDst);
    if (FLT_EPSILON < dist) {
        dirToForcePosX = offsetX / dist;
        dirToForcePosY = offsetY / dist;
    }
    float centreT = 1 - dist / p.radius;
    float velX = vel[particleIndex][0];
    float velY = vel[particleIndex][1];
    outForce[0] = (dirToForcePosX * p.strength - velX) * centreT;
    outForce[1] = (dirToForcePosY * p.strength - velY) * centreT;
}

std::vector<int> World::queryNeighbors(float x, float y) const {
    auto candidates = gridmap.findNeighborhood(x, y, smoothingRadius);
    std::vector<int> result;
    result.reserve(candidates.size());
    float r2 = smoothingRadius * smoothingRadius;
    for (int idx : candidates) {
        float dx = pos[idx][0] - x;
        float dy = pos[idx][1] - y;
        if (dx * dx + dy * dy <= r2) {
            result.push_back(idx);
        }
    }
    return result;
}

std::vector<int> World::querySpatialHash(float x, float y) const {
    return gridmap.findNeighborhood(x, y, smoothingRadius);
}

World::~World() {
#ifdef SPH_ENABLE_HASH2D
    freeDeviceBuffers();
#endif
}

#ifdef SPH_ENABLE_HASH2D
void World::allocateDeviceBuffers() {
    if (device_allocated) return;
    uint32_t N = static_cast<uint32_t>(numParticle);
    grid.gridDim = make_uint2(
        static_cast<uint32_t>(std::ceil(worldSize[0] / smoothingRadius)) + 1,
        static_cast<uint32_t>(std::ceil(worldSize[1] / smoothingRadius)) + 1);
    grid.invCell = 1.0f / smoothingRadius;
    grid.gridCells = grid.gridDim.x * grid.gridDim.y;
    CUDA_TRY(cudaMalloc(&grid.particles.pos, N * sizeof(float2)));
    CUDA_TRY(cudaMalloc(&grid.hashBuf, N * sizeof(uint32_t)));
    CUDA_TRY(cudaMalloc(&grid.idxBuf, N * sizeof(uint32_t)));
    CUDA_TRY(cudaMalloc(&grid.cellStart, grid.gridCells * sizeof(uint32_t)));
    CUDA_TRY(cudaMalloc(&grid.cellEnd, grid.gridCells * sizeof(uint32_t)));
    CUDA_TRY(cudaMalloc(&d_neighbors, N * MAX_NEIGHBORS * sizeof(uint32_t)));
    CUDA_TRY(cudaMalloc(&d_counts, N * sizeof(uint32_t)));
    device_allocated = true;
}

void World::freeDeviceBuffers() {
    if (!device_allocated) return;
    CUDA_TRY(cudaFree(grid.particles.pos));
    CUDA_TRY(cudaFree(grid.hashBuf));
    CUDA_TRY(cudaFree(grid.idxBuf));
    CUDA_TRY(cudaFree(grid.cellStart));
    CUDA_TRY(cudaFree(grid.cellEnd));
    CUDA_TRY(cudaFree(d_neighbors));
    CUDA_TRY(cudaFree(d_counts));
    device_allocated = false;
}
#endif

} // namespace sph

File: sph-core/src/world.rs
use crate::config::WorldConfig;
use crate::grid_map::GridMap;
use crate::kernels::{calc_smoothing_kernel, calc_smoothing_kernel_derivative};
#[cfg(feature = "cuda")]
use sph_cuda;

#[derive(Debug, Clone)]
pub struct ForcePoint {
    pub pos: [f32; 2],
    pub radius: f32,
    pub strength: f32,
}

#[derive(Debug)]
pub struct World {
    _particle_radius: i32,
    num_particle: usize,
    gravity: f32,
    world_size: [f32; 2],
    collision_damping: f32,
    smoothing_radius: f32,
    target_density: f32,
    pressure_multiplier: f32,
    delta: f32,
    drag: f32,
    mass: Vec<f32>,

    pos: Vec<[f32; 2]>,
    predpos: Vec<[f32; 2]>,
    vel: Vec<[f32; 2]>,
    density: Vec<f32>,
    pressure_accelerations: Vec<[f32; 2]>,
    interaction_force: Vec<[f32; 2]>,
    color: Vec<[u8; 3]>,
    querysize: Vec<Vec<usize>>,

    force_point: ForcePoint,
    gridmap: GridMap,
}

impl World {
    pub const DEFAULT_NUM_PARTICLES: usize = 1_000;

    pub fn new(config: WorldConfig) -> Self {
        let num_particle = config.num_particles.max(1);
        let world_size = [config.world_width, config.world_height];
        let mut world = Self {
            _particle_radius: 5,
            num_particle,
            gravity: config.gravity,
            world_size,
            collision_damping: config.collision_damping,
            smoothing_radius: config.smoothing_radius,
            target_density: config.target_density,
            pressure_multiplier: config.pressure_multiplier,
            delta: config.delta,
            drag: config.drag,
            mass: vec![1.0; num_particle],
            pos: Vec::with_capacity(num_particle),
            predpos: vec![[0.0; 2]; num_particle],
            vel: vec![[0.0; 2]; num_particle],
            density: vec![0.0; num_particle],
            pressure_accelerations: vec![[0.0; 2]; num_particle],
            interaction_force: vec![[0.0; 2]; num_particle],
            color: vec![[255, 255, 255]; num_particle],
            querysize: vec![Vec::new(); num_particle],
            force_point: ForcePoint {
                pos: [0.0, 0.0],
                radius: 0.0,
                strength: 0.0,
            },
            gridmap: GridMap::new(world_size[0], world_size[1], config.smoothing_radius),
        };

        world.initialize_positions();
        world
    }

    fn initialize_positions(&mut self) {
        let side = (self.num_particle as f32).sqrt().ceil() as usize;
        for i in 0..self.num_particle {
            let row = i / side;
            let col = i % side;
            let x = (col as f32 / side as f32) * self.world_size[0];
            let y = (row as f32 / side as f32) * self.world_size[1];
            self.pos.push([x, y]);
        }
    }

    pub fn step(&mut self, delta_time: f32) {
        self.update(delta_time, false);
    }

    pub fn step_with_cuda(&mut self, delta_time: f32) {
        self.update(delta_time, true);
    }

    pub fn set_interaction_force(&mut self, pos_x: f32, pos_y: f32, radius: f32, strength: f32) {
        self.force_point = ForcePoint {
            pos: [pos_x, pos_y],
            radius,
            strength,
        };
    }

    pub fn delete_interaction_force(&mut self) {
        self.force_point = ForcePoint {
            pos: [0.0, 0.0],
            radius: 0.0,
            strength: 0.0,
        };
    }

    pub fn get_world_width(&self) -> f32 {
        self.world_size[0]
    }

    pub fn get_world_height(&self) -> f32 {
        self.world_size[1]
    }

    pub fn get_smoothing_radius(&self) -> f32 {
        self.smoothing_radius
    }

    pub fn get_gravity(&self) -> f32 {
        self.gravity
    }

    pub fn get_drag(&self) -> f32 {
        self.drag
    }

    pub fn get_target_density(&self) -> f32 {
        self.target_density
    }

    pub fn get_pressure_multiplier(&self) -> f32 {
        self.pressure_multiplier
    }

    pub fn get_delta(&self) -> f32 {
        self.delta
    }

    pub fn get_collision_damping(&self) -> f32 {
        self.collision_damping
    }

    pub fn get_num_particles(&self) -> usize {
        self.num_particle
    }

    pub fn positions(&self) -> &[[f32; 2]] {
        &self.pos
    }

    pub fn velocities(&self) -> &[[f32; 2]] {
        &self.vel
    }

    pub fn query_neighbors(&self, x: f32, y: f32) -> Vec<usize> {
        let candidates = self.gridmap.find_neighborhood(x, y, self.smoothing_radius);
        let r2 = self.smoothing_radius * self.smoothing_radius;
        candidates
            .into_iter()
            .filter(|idx| {
                let dx = self.pos[*idx][0] - x;
                let dy = self.pos[*idx][1] - y;
                dx * dx + dy * dy <= r2
            })
            .collect()
    }

    pub fn query_spatial_hash(&self, x: f32, y: f32) -> Vec<usize> {
        self.gridmap.find_neighborhood(x, y, self.smoothing_radius)
    }

    fn update(&mut self, delta_time: f32, prefer_cuda: bool) {
        self.predicted_pos(delta_time);
        self.gridmap.unregister_all();

        for idx in 0..self.num_particle {
            let p = self.pos[idx];
            self.gridmap.register_target(idx, p[0], p[1]);
        }

        self.querysize = (0..self.num_particle)
            .map(|idx| {
                let p = self.pos[idx];
                self.gridmap
                    .find_neighborhood(p[0], p[1], self.smoothing_radius)
            })
            .collect();

        for idx in 0..self.num_particle {
            self.update_density(idx, prefer_cuda);
        }
        for idx in 0..self.num_particle {
            self.update_pressure_force(idx);
        }
        for idx in 0..self.num_particle {
            self.update_interaction_force(idx);
        }

        self.update_position(delta_time);
        for idx in 0..self.num_particle {
            self.fix_position_from_world_size(idx);
        }
        self.update_color();
    }

    fn predicted_pos(&mut self, delta_time: f32) {
        for i in 0..self.num_particle {
            self.vel[i][1] += self.mass[i] * self.gravity * delta_time;
            self.predpos[i][0] = self.pos[i][0] + self.vel[i][0] * (1.0 / 120.0);
            self.predpos[i][1] = self.pos[i][1] + self.vel[i][1] * (1.0 / 120.0);
        }
    }

    fn update_density(&mut self, particle_index: usize, prefer_cuda: bool) {
        self.density[particle_index] = self.calc_density(particle_index, prefer_cuda);
    }

    fn update_pressure_force(&mut self, particle_index: usize) {
        let mut pressure_force = [0.0, 0.0];
        self.calc_pressure_force(&mut pressure_force, particle_index);
        let divisor = self.density[particle_index] + self.delta;
        self.pressure_accelerations[particle_index][0] = pressure_force[0] / divisor;
        self.pressure_accelerations[particle_index][1] = pressure_force[1] / divisor;
    }

    fn update_interaction_force(&mut self, particle_index: usize) {
        let mut force = [0.0, 0.0];
        self.calc_interaction_force(&mut force, particle_index);
        self.interaction_force[particle_index] = force;
    }

    fn update_position(&mut self, delta_time: f32) {
        for i in 0..self.num_particle {
            self.vel[i][0] +=
                (self.pressure_accelerations[i][0] + self.interaction_force[i][0]) * delta_time;
            self.vel[i][1] +=
                (self.pressure_accelerations[i][1] + self.interaction_force[i][1]) * delta_time;
            self.pos[i][0] += self.vel[i][0] * delta_time;
            self.pos[i][1] += self.vel[i][1] * delta_time;
            self.vel[i][0] *= self.drag;
            self.vel[i][1] *= self.drag;
        }
    }

    fn fix_position_from_world_size(&mut self, i: usize) {
        let x = self.pos[i][0];
        let y = self.pos[i][1];
        let vel_x = self.vel[i][0];
        let vel_y = self.vel[i][1];
        let w = self.world_size[0];
        let h = self.world_size[1];
        if x < 0.0 {
            self.pos[i][0] = 0.0;
            self.vel[i][0] = -vel_x * self.collision_damping;
        }
        if w < x {
            self.pos[i][0] = w;
            self.vel[i][0] = -vel_x * self.collision_damping;
        }
        if y < 0.0 {
            self.pos[i][1] = 0.0;
            self.vel[i][1] = -vel_y * self.collision_damping;
        }
        if h < y {
            self.pos[i][1] = h;
            self.vel[i][1] = -vel_y * self.collision_damping;
        }
    }

    fn update_color(&mut self) {
        let mut speeds = Vec::with_capacity(self.num_particle);
        let mut min_speed = f32::MAX;
        let mut max_speed = 0.0_f32;
        for i in 0..self.num_particle {
            let speed = (self.vel[i][0] * self.vel[i][0] + self.vel[i][1] * self.vel[i][1]).sqrt();
            min_speed = min_speed.min(speed);
            max_speed = max_speed.max(speed);
            speeds.push(speed);
        }

        for (i, speed) in speeds.into_iter().enumerate() {
            let denom = (max_speed - min_speed).max(f32::EPSILON);
            let norm_speed = ((speed - min_speed) / denom).clamp(0.0, 1.0);
            let byte_vel = (norm_speed * 255.0) as u8;
            let color1 = [0, 0, 255];
            let color2 = [255, 0, 0];
            self.color[i][0] = (((255 - byte_vel) as u16 * color1[0] as u16
                + byte_vel as u16 * color2[0] as u16)
                / 255) as u8;
            self.color[i][1] = (((255 - byte_vel) as u16 * color1[1] as u16
                + byte_vel as u16 * color2[1] as u16)
                / 255) as u8;
            self.color[i][2] = (((255 - byte_vel) as u16 * color1[2] as u16
                + byte_vel as u16 * color2[2] as u16)
                / 255) as u8;
        }
    }

    fn calc_density(&self, particle_index: usize, prefer_cuda: bool) -> f32 {
        #[cfg(not(feature = "cuda"))]
        let _ = prefer_cuda;

        let other_indexes = &self.querysize[particle_index];
        if other_indexes.is_empty() {
            return 0.0;
        }

        let distances: Vec<f32> = other_indexes
            .iter()
            .map(|&j| {
                let dx = self.predpos[j][0] - self.predpos[particle_index][0];
                let dy = self.predpos[j][1] - self.predpos[particle_index][1];
                (dx * dx + dy * dy).sqrt()
            })
            .collect();

        #[cfg(feature = "cuda")]
        if prefer_cuda {
            if let Ok(values) = sph_cuda::calc_smoothing_kernel(&distances, self.smoothing_radius) {
                return other_indexes
                    .iter()
                    .zip(values.into_iter())
                    .map(|(&j, val)| self.mass[j] * val)
                    .sum();
            }
        }

        other_indexes
            .iter()
            .zip(distances.iter())
            .map(|(&j, &dist)| self.mass[j] * calc_smoothing_kernel(dist, self.smoothing_radius))
            .sum()
    }

    fn calc_pressure_force(&self, pressure_force: &mut [f32; 2], particle_index: usize) {
        for &other_index in &self.querysize[particle_index] {
            if particle_index == other_index {
                continue;
            }
            let offset_x = self.pos[other_index][0] - self.pos[particle_index][0];
            let offset_y = self.pos[other_index][1] - self.pos[particle_index][1];
            let dist = (offset_x * offset_x + offset_y * offset_y).sqrt();
            if dist > self.smoothing_radius {
                continue;
            }
            let (dir_x, dir_y) = if dist <= f32::EPSILON {
                (1.0_f32, 0.0_f32)
            } else {
                (offset_x / dist, offset_y / dist)
            };
            let slope = calc_smoothing_kernel_derivative(dist, self.smoothing_radius);
            let other_density = self.density[other_index];
            let shared_pressure =
                self.calc_shared_pressure(other_density, self.density[particle_index]);
            let a = shared_pressure * slope * self.mass[other_index] / (other_density + self.delta);
            pressure_force[0] += dir_x * a;
            pressure_force[1] += dir_y * a;
        }
    }

    fn calc_shared_pressure(&self, density_left: f32, density_right: f32) -> f32 {
        let pressure_left = self.convert_density_to_pressure(density_left);
        let pressure_right = self.convert_density_to_pressure(density_right);
        (pressure_left + pressure_right) * 0.5
    }

    fn convert_density_to_pressure(&self, density_val: f32) -> f32 {
        let density_error = density_val - self.target_density;
        density_error * self.pressure_multiplier
    }

    fn calc_interaction_force(&self, out_force: &mut [f32; 2], particle_index: usize) {
        out_force[0] = 0.0;
        out_force[1] = 0.0;
        let p = &self.force_point;
        let offset_x = p.pos[0] - self.pos[particle_index][0];
        let offset_y = p.pos[1] - self.pos[particle_index][1];
        let sqr_dst = offset_x * offset_x + offset_y * offset_y;
        if sqr_dst >= p.radius * p.radius {
            return;
        }
        let dist = sqr_dst.sqrt();
        let (dir_to_force_pos_x, dir_to_force_pos_y) = if dist > f32::EPSILON {
            (offset_x / dist, offset_y / dist)
        } else {
            (0.0, 0.0)
        };
        let centre_t = 1.0 - dist / p.radius;
        let vel_x = self.vel[particle_index][0];
        let vel_y = self.vel[particle_index][1];
        out_force[0] = (dir_to_force_pos_x * p.strength - vel_x) * centre_t;
        out_force[1] = (dir_to_force_pos_y * p.strength - vel_y) * centre_t;
    }
}
